### ISSUE TITLE
Inline comments in types

### ISSUE THREAD

**Issue by AndrolGenhald** [vimeo/psalm#8168]
This has come up on slack once or twice, and I've found myself wanting it a number of times. Something like this would be very useful to help document types:

```
/**
 * @return array<
 *     array-key, --id (TODO change to int)--
 *     string, --name--
 * >
 */
```

A couple things to note about syntax:

 - We should make sure that whatever syntax we use allows every character in the comment, we shouldn't disallow comments like `TODO change to iterable<int, string>` for instance (eg if the syntax for comments was everything after the first space and before the next comma this would fail)
 - Comments should ideally be allowed in single line types, with multiple comments in a single line

I think using `--[comment]--` would work ok, but I haven't thought too much about it and I'm very open to alternatives. One plus about using hyphens is that you can't use them in class names, and internal types never use double hyphens, so it should always remain unambiguous.

**Comment by kkmuffme** [comment]
WordPress does this with {} notation after the type. and I think having comments OUTSIDE of the actual param types is preferable. (even then I'm not a big fan of that)

By adding support for comments within the types, we will break literally all other phpdoc parsers.

e.g

``` 
@param string|array $args {
     *     Title attribute arguments. Optional.
     *
     *     @type string  $before Markup to prepend to the title. Default empty.
     *     @type string  $after  Markup to append to the title. Default empty.
     *     @type bool    $echo   Whether to echo or return the title. Default true for echo.
     *     @type WP_Post $post   Current post object to retrieve the title for.
     * }
```

**Comment by AndrolGenhald** [comment]
> WordPress does this with {} notation after the type. and I think having comments OUTSIDE of the actual param types is preferable. (even then I'm not a big fan of that)

That doesn't really work well for deeply-nested types, compare:

```php
/**
 * Maps country, state/province/territory/etc, city to population
 * @return array<string, array<string, array<string, int<0, max>>>>
 */
```

```php
/**
 * Maps country, state/province/territory/etc, city to population
 * @return array<
 *     string, --country--
 *     array<
 *         string, --state/province/territory/etc--
 *         array<
 *             string, --city--
 *             int<0, max> --population--
 *         >
 *     >
 * >
 */
```

Some of these sorts of things can get pretty confusing at first glance, having inline comments could help significantly.

> By adding support for comments within the types, we will break literally all other phpdoc parsers.

Yeah, we should probably discuss this with the wider community and settle on a standard syntax.

**Comment by kkmuffme** [comment]
>That doesn't really work well for deeply-nested types, compare:

I know, I'm just saying - WP has been doing this for more than a decade now, and it hasn't been adopted by any other project nor is it supported by any parser (well, there is a custom WP hooks parser that supports it)
I think those cmoments are a dead horse.
Given the above "city/population/... example, I think it would be better to actually make this a named array{city: string, ...}

**Comment by AndrolGenhald** [comment]
> I know, I'm just saying - WP has been doing this for more than a decade now, and it hasn't been adopted by any other project nor is it supported by any parser (well, there is a custom WP hooks parser that supports it)

:+1: 

I think that works well if you're not using array shapes, but if you want to have Psalm understand the array shape type this seems excessively verbose and less readable:

```
 * @param string|array{
 *     before: string,
 *     after:  string,
 *     echo:   bool,
 *     post:   WP_Post,
 * } $args {
 *     Title attribute arguments. Optional.
 *
 *     @type string  $before Markup to prepend to the title. Default empty.
 *     @type string  $after  Markup to append to the title. Default empty.
 *     @type bool    $echo   Whether to echo or return the title. Default true for echo.
 *     @type WP_Post $post   Current post object to retrieve the title for.
 * }
```

vs

```
 * @param string|array{
 *     before: string,  --Markup to prepend to the title. Default empty.--
 *     after:  string,  --Markup to append to the title. Default empty.--
 *     echo:   bool,    --Whether to echo or return the title. Default true for echo.--
 *     post:   WP_Post, --Current post object to retrieve the title for.--
 * } $args Title attribute arguments. Optional.
```

> Given the above "city/population/... example, I think it would be better to actually make this a named array{city: string, ...}

I've got a project with a lot of types like this, and they must be maps, they need to be accessed like `$population = $population_map[$country][$state][$city]` for example. Doing a full scan of the array isn't feasible for performance.

**Comment by mckravchyk** [comment]
I don't think that inline comments would do. It's certainly better than nothing but in my experience the properties that need to be commented the most contain multiple sentences and take multiple lines. If a property can be described in a few words, it often does not need a comment in the first place. Obviously lines that stick out to too much to the right of the screen (beyond the max length limit set in the linter config)  are a big no no when it comes to code style.

**Comment by AndrolGenhald** [comment]
@mckravchyk I would think inline comments could serve just as well as multi-line or separate-line comments, but separate-line comments definitely could not be used for inline comments.

**Comment by weirdan** [comment]
Implemented in #10004 ([Psalm 5.14.0](https://github.com/vimeo/psalm/releases/tag/5.14.0))
